### PR TITLE
Transform paragraph block to buttons block

### LINKED ISSUE #28298: Allow Paragraph Block to Transform to a Buttons Block
## What problem does this address?

Especially when transforming content from the Classic Editor or other CMSes to blocks, I am often faced with a paragraph that contains only a link and should be a button.

## What is your proposed solution?
If a paragraph only contains a link, allow it to be transformed to a button with the existing link and anchor text.

Diagram of the transform. A paragraph block with "Link in Paragraph" link text followed by a transformed button block with "Link in Paragraph" label

Screenshot of Paragraph block transform options with "Button" written in at end of transform options

### PR DESCRIPTION
**Pull request by hsingyuc**
## Description
Fix #28298 
Transform paragraph block to buttons block

## How has this been tested?
// Test text with one link
1. Create text ( ex. Link in Paragraph ) with a link in a paragraph block
2. Mark `Link in Paragraph`, click `Change block type or style` and see if `Buttons` renders as expected.
3. Click `Buttons` and see if the text transforms into a button with the link.

// Test texts with links
1. Create text ( ex. Link in Paragraph ) with links ( ex. link1  link2  link3 ) in a paragraph block
2. Mark `Link in Paragraph`, click `Change block type or style` and see if `Buttons` renders as expected.
3. Click `Buttons` and see if the text transforms into a button with the first link (link1).

// Test texts without a link
1. Create text ( ex. Link in Paragraph ) without a link in a paragraph block
2. Mark `Link in Paragraph`, click `Change block type or style` and see if `Buttons` renders as expected.
3. Click `Buttons` and see if the text transforms into a button.

## Screenshots 
1. Before
before
paragraph

2. After
after
button

## Checklist:
- [X ] My code is tested.
- [X] My code follows the WordPress code style.

### CODE REVIEW

**Comment by mtias** [comment]
Side note: one thing we need to revise a bit is the order of transformations. Headings and List should probably come first, for example. We might also want to restrict certain transforms to the case where the length of a paragraph block is < than some characters — it doesn't make sense to convert a full paragraph with multiple lines into a button, for example.

**Comment by ntsekouras** [comment]
> We might also want to restrict certain transforms

Maybe also checking if a paragraph has more than one links?

**Comment by hsingyuc** [comment]
>Headings and List should probably come first

It renders based on the usage so I think this is the current expected behavior.  If we want to change, this may be a follow-up issue.
螢幕快照 2021-01-27 下午5 08 11

>We might also want to restrict certain transforms to the case where the length of a paragraph block is < than some characters 

I didn't know there was a way to do this.  Any suggestions on checking length before showing the transform option in the transforms list?

>Maybe also checking if a paragraph has more than one links?

Do you mean if more than one link then it transforms to multiple buttons or if more than a link then do not transform?

**Comment by ntsekouras** [comment]
>It renders based on the usage so I think this is the current expected behavior

Yes, I think this is out of scope for this PR.

>I didn't know there was a way to do this. Any suggestions on checking length before showing the transform option in the transforms list?

You can use `isMatch` function from Block Transforms API: https://developer.wordpress.org/block-editor/developers/block-api/block-transforms/#block

>Do you mean if more than one link then it transforms to multiple buttons or if more than a link then do not transform?

I was thinking of not transforming.

**Comment by hsingyuc** [comment]
@mtias  @ntsekouras  Thank you for the feedback! The changes have been pushed. ✌️

File: packages/block-library/src/buttons/transforms.js [hunk 1]
```diff
@@ -26,6 +26,42 @@ const transforms = {
...
+					{},
+					// Loop the selected buttons
+					buttons.map( ( attributes ) => {
+						// Remove any HTML tags
+						const div = document.createElement( 'div' );
+						div.innerHTML = attributes.content;
+						const text = div.innerText || '';
+						// Get first url
+						const link = div.querySelector( 'a' );
+						const url = link ? link.getAttribute( 'href' ) : null;
```
Comment (ntsekouras) [inline packages/block-library/src/buttons/transforms.js:46]:
```suggestion
						const url = link?.getAttribute( 'href' );
```

File: packages/block-library/src/buttons/transforms.js [hunk 2]
```diff
@@ -26,6 +26,42 @@ const transforms = {
...
+						// Create singular button in the buttons block
+						return createBlock( 'core/button', {
+							text,
+							url,
+						} );
+					} )
+				),
+			isMatch: ( paragraphs ) => {
+				const buttons = paragraphs.filter( ( attributes ) => {
```
Comment (ntsekouras) [inline packages/block-library/src/buttons/transforms.js:55]: You can use [Array.prototype.every()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) here.

**Review by ntsekouras — COMMENTED** [review]
Good work here @hsingyuc 👍 - thanks for working on this!

I left a couple minor comments/nitpicks, but your code works as expected.

The maximum length for the paragraph to **thirty** (30) seems a good choice to me, but would like some more feedback about that (@mtias )

**Comment by mtias** [comment]
> The maximum length for the paragraph to thirty (30) seems a good choice to me...

Let's try it, we can always adjust :)

**Comment by hsingyuc** [comment]
@ntsekouras  Thanks for the tips!! I made the changes.

**Review by ntsekouras — APPROVED** [review]
Thank you! 👍

**Comment by mtias** [comment]
Thanks for working on this @hsingyuc !

**Pull request merged**